### PR TITLE
Fix sampler state propagation and cache refresh during pretraining

### PR DESCRIPTION
**Pull request by franknoe**
## Summary

Fix sampler state handling in orbital pretraining without changing its loss,
optimizer, RNG sequence, or public generator output.

The production-code diff is 8 added lines and 2 removed lines. A separate
94-line regression test covers three cases.

## Problem

`pretrain_step()` receives the state returned by `sampler.sample()`, but does
not return it to the outer loop. Consequently each iteration starts from the
original sampler state rather than the advanced walkers/adaptation state.

Parameters also change at each optimizer step. The sampler's cached
wavefunction values are not refreshed, so subsequent Metropolis acceptance
ratios can compare proposals evaluated with current parameters against values
cached using old parameters. Langevin sampling additionally caches forces.

## Fix

- Return and retain the advanced sampler state.
- Refresh it through `sampler.update()` **after** `pmap_merge_states()` so caches
  match the final parameters, mirroring the existing `fit_wf()` training path.
- Preserve the four-element public yield and the existing orbital-MSE objective.

## Tests

Self-contained tests exercise the real mapped pretraining loop with a simple
sampler/target and analytically predictable losses:

1. Sampler state advances between iterations.
2. Cached values are refreshed after optimizer updates.
3. Refresh happens after parameter merging.

All three fail on unmodified upstream `edf373e7` and pass with this patch.

Validation on macOS/arm64, Python 3.12, JAX 0.8.3:

- `PYTHONPATH=src python -m pytest tests/test_pretraining.py tests/test_sampling.py -q`: **13 passed**.
- `PYTHONPATH=src XLA_FLAGS=--xla_force_host_platform_device_count=2 python -m pytest tests/test_pretraining.py -q`: **3 passed** on two CPU devices.
- Upstream pre-commit checks for both changed files: passed.
- Pyright on the changed production module: **0 errors, 0 warnings**.

No experiment scripts, numerical data, or unrelated refactoring are included.
The full upstream test suite was not run.

### CODE REVIEW

**Comment by zenoone** [comment]
Thanks for the clean PR. Yes, this indeed looks like a genuine bug that we haven't spotted! For context, in our recent workflows we usually did not employ the pre-training. Of course, it got introduced with a PR having a [WIP] label.

**Review by zenoone — APPROVED** [review]
